test-runner/src/main/scala/scala/scalanative/testinterface/ComRunner.scala

```diff
@@ -1,112 +1,203 @@
+                  case _: EOFException => close()
+                }
+            }
+          }
+        } catch {
+          // We got interrupted by a graceful close. This is OK.
+          case _: IOException if state == Closing => ()
+        }
+
+        /**
+         * Everything got closed. We wait for the run to terminate.
+         * We need to wait in order to make sure that closing the
+         * underlying run does not fail it. */
+        processRunner.future.foreach { _ =>
+          processRunner.close()
+          promise.trySuccess(())
+        }
+      } catch {
+        case t: Throwable => handleThrowable(t)
+      }
     }
-  }
+  }.start()
 
-  private[this] var serverSocket: ServerSocket = _
-  private[this] val socket =
-    try {
-      serverSocket = new ServerSocket( /* port = */ 0, /* backlog = */ 1)
+  val future: Future[Unit] = promise.future
 
-      runner.start()
+  def send(msg: String): Unit = synchronized {
+    state match {
+      case AwaitingConnection(msgs) =>
+        state = AwaitingConnection(msg :: msgs)
 
-      serverSocket.setSoTimeout(40 * 1000)
-      serverSocket.accept()
-    } catch {
-      case _: SocketTimeoutException =>
-        throw new BuildException(
-          "The test program never connected to the test runner.")
-    } finally {
-      // We can close it immediately, since we won't receive another connection.
-      serverSocket.close()
+      case Connected(_, jvm2native, _) =>
+        try {
+          writeMsg(jvm2native, msg)
+          jvm2native.flush()
+        } catch {
+          case t: Throwable => handleThrowable(t)
+        }
+
+      case Closing => () // ignore msg.
     }
+  }
 
-  private[this] val in = new DataInputStream(
-    new BufferedInputStream(socket.getInputStream))
-  private[this] val out = new DataOutputStream(
-    new BufferedOutputStream(socket.getOutputStream))
-
-  /** Send message `msg` to the distant program. */
-  def send(msg: Message): Unit = synchronized {
-    try SerializedOutputStream(out)(_.writeMessage(msg))
-    catch {
-      case ex: Throwable =>
-        close()
-        throw ex
+  def close(): Unit = synchronized {
+    val oldState = state
+    state = Closing // Signal receiver thread that it is OK if socket read fails.
+    oldState match {
+      case c: Connected =>
+        closeAll(
+          c
+        ) // Interrupts the receiver thread and signals the VM to terminate.
+      case Closing | _: AwaitingConnection => ()
     }
   }
 
-  /** Wait for a message to arrive from the distant program. */
-  def receive(): Message = synchronized {
+  private def onNativeTerminated(): Unit = {
+    close()
+
+    /**
+     * Interrupt receiver if we are still waiting for connection.
+     * Should only be relevant if we are still awaiting the connection.
+     * Note: We cannot do this in close(), otherwise if the JVM side closes
+     * before the JS side connected, the JS VM will fail instead of terminate
```

Comment:
`JS side`  holdover from port?   I am still studying the larger communications context.

Reply:
Yes, I've missed to fix this (and few others). It's fixed now